# Content
Analysis of hotel data which contains booking information for a city hotel and a resort hotel, and includes information such as when the booking was made, length of stay, the number of adults, children, and/or babies, and the number of available parking spaces, among other things.

This data set contains a single file which compares various booking information between two hotels: a city hotel and a resort hotel.

>You will Find the data used on kaggle [here](https://www.kaggle.com/datasets/jessemostipak/hotel-booking-demand).


# Some Feature Description

1-is_canceled : Value indicating if the booking was canceled (1) or not (0) - Target feature


2-hotel : Hotel (H1 = Resort Hotel or H2 = City Hotel)

3-lead_time : Number of days that elapsed between the entering date of the booking into the PMS and the arrival date

4-is_repeated_guest : Value indicating if the booking name was from a repeated guest (1) or not (0)


5-previous_cancellations : Number of previous bookings that were cancelled by the customer prior to the current booking


6-previous_bookings_not_canceled : Number of previous bookings not cancelled by the customer prior to the current booking


7-reserved_room_type : Code of room type reserved. Code is presented instead of designation for anonymity reasons.


8-assigned_room_type : Code for the type of room assigned to the booking. Sometimes the assigned room type differs from the reserved room type due to hotel operation reasons (e.g. overbooking) or by customer request. Code is presented instead of designation for anonymity reasons.

9-booking_changes : Number of changes/amendments made to the booking from the moment the booking was entered on the PMS until the moment of check-in or cancellation

10-deposit_type : Indication on if the customer made a deposit to guarantee the booking. This variable can assume three categories: No Deposit – no deposit was made; Non Refund – a deposit was made in the value of the total stay cost; Refundable – a deposit was made with a value under the total cost of stay.

In [1]:
# import important libraries
import numpy as np # linear algebra
import pandas as pd #data processing,  read CSV file 

In [2]:
df=pd.read_csv(r"C:\Users\otaku7\Downloads\archive (2)\hotel_bookings.csv")

In [3]:
df.head()

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   
2  Resort Hotel            0          7               2015               July   
3  Resort Hotel            0         13               2015               July   
4  Resort Hotel            0         14               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  ...  deposit_type  \
0                        0                     0       2  ...    No Deposit   
1                        0                     0       2  ...    No Deposit   
2                        0                     1       1  ...    No Deposit   
3                        0                     1       1  ...    No Deposit   
4                        0                     2       2  ...    No Deposit   

   agent company days_in_waiting_list customer_type   adr  \
0    NaN     NaN                    0     Transient   0.0   
1    NaN     NaN                    0     Transient   0.0   
2    NaN     NaN                    0     Transient  75.0   
3  304.0     NaN                    0     Transient  75.0   
4  240.0     NaN                    0     Transient  98.0   

   required_car_parking_spaces  total_of_special_requests  reservation_status  \
0                            0                          0           Check-Out   
1                            0                          0           Check-Out   
2                            0                          0           Check-Out   
3                            0                          0           Check-Out   
4                            0                          1           Check-Out   

  reservation_status_date  
0              2015-07-01  
1              2015-07-01  
2              2015-07-02  
3              2015-07-02  
4              2015-07-03  

[5 rows x 32 columns]

# Understanding of data


In [4]:
df.shape

(119390, 32)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

This data contains 4 columns of float type , 16 columns of int type and 12 column of object type

In [6]:
# Viewing column names
df.columns

Index(['hotel', 'is_canceled', 'lead_time', 'arrival_date_year',
       'arrival_date_month', 'arrival_date_week_number',
       'arrival_date_day_of_month', 'stays_in_weekend_nights',
       'stays_in_week_nights', 'adults', 'children', 'babies', 'meal',
       'country', 'market_segment', 'distribution_channel',
       'is_repeated_guest', 'previous_cancellations',
       'previous_bookings_not_canceled', 'reserved_room_type',
       'assigned_room_type', 'booking_changes', 'deposit_type', 'agent',
       'company', 'days_in_waiting_list', 'customer_type', 'adr',
       'required_car_parking_spaces', 'total_of_special_requests',
       'reservation_status', 'reservation_status_date'],
      dtype='object')

In [7]:
#statisical details of all the columns
df.describe(include="all")

hotel    is_canceled      lead_time  arrival_date_year  \
count       119390  119390.000000  119390.000000      119390.000000   
unique           2            NaN            NaN                NaN   
top     City Hotel            NaN            NaN                NaN   
freq         79330            NaN            NaN                NaN   
mean           NaN       0.370416     104.011416        2016.156554   
std            NaN       0.482918     106.863097           0.707476   
min            NaN       0.000000       0.000000        2015.000000   
25%            NaN       0.000000      18.000000        2016.000000   
50%            NaN       0.000000      69.000000        2016.000000   
75%            NaN       1.000000     160.000000        2017.000000   
max            NaN       1.000000     737.000000        2017.000000   

       arrival_date_month  arrival_date_week_number  \
count              119390             119390.000000   
unique                 12                       NaN   
top                August                       NaN   
freq                13877                       NaN   
mean                  NaN                 27.165173   
std                   NaN                 13.605138   
min                   NaN                  1.000000   
25%                   NaN                 16.000000   
50%                   NaN                 28.000000   
75%                   NaN                 38.000000   
max                   NaN                 53.000000   

        arrival_date_day_of_month  stays_in_weekend_nights  \
count               119390.000000            119390.000000   
unique                        NaN                      NaN   
top                           NaN                      NaN   
freq                          NaN                      NaN   
mean                    15.798241                 0.927599   
std                      8.780829                 0.998613   
min                      1.000000                 0.000000   
25%                      8.000000                 0.000000   
50%                     16.000000                 1.000000   
75%                     23.000000                 2.000000   
max                     31.000000                19.000000   

        stays_in_week_nights         adults  ...  deposit_type          agent  \
count          119390.000000  119390.000000  ...        119390  103050.000000   
unique                   NaN            NaN  ...             3            NaN   
top                      NaN            NaN  ...    No Deposit            NaN   
freq                     NaN            NaN  ...        104641            NaN   
mean                2.500302       1.856403  ...           NaN      86.693382   
std                 1.908286       0.579261  ...           NaN     110.774548   
min                 0.000000       0.000000  ...           NaN       1.000000   
25%                 1.000000       2.000000  ...           NaN       9.000000   
50%                 2.000000       2.000000  ...           NaN      14.000000   
75%                 3.000000       2.000000  ...           NaN     229.000000   
max                50.000000      55.000000  ...           NaN     535.000000   

            company days_in_waiting_list customer_type            adr  \
count   6797.000000        119390.000000        119390  119390.000000   
unique          NaN                  NaN             4            NaN   
top             NaN                  NaN     Transient            NaN   
freq            NaN                  NaN         89613            NaN   
mean     189.266735             2.321149           NaN     101.831122   
std      131.655015            17.594721           NaN      50.535790   
min        6.000000             0.000000           NaN      -6.380000   
25%       62.000000             0.000000           NaN      69.290000   
50%      179.000000             0.000000           NaN      94.575000   
75%      270.000000             0.000000           

# Cleaning of data

In [8]:
# checking missing values count in the data
df.isnull().sum()

hotel                                  0
is_canceled                            0
lead_time                              0
arrival_date_year                      0
arrival_date_month                     0
arrival_date_week_number               0
arrival_date_day_of_month              0
stays_in_weekend_nights                0
stays_in_week_nights                   0
adults                                 0
children                               4
babies                                 0
meal                                   0
country                              488
market_segment                         0
distribution_channel                   0
is_repeated_guest                      0
previous_cancellations                 0
previous_bookings_not_canceled         0
reserved_room_type                     0
assigned_room_type                     0
booking_changes                        0
deposit_type                           0
agent                              16340
company         

4 columns contains missing values, but the The largest column contains missing values is company (no of total rows is 119390)so i will drop this column 


In [9]:
df.drop("company",axis=1,inplace=True)

In [10]:
df.shape

(119390, 31)

another way to check if columns have many of missing values is computing missing values % in these columns 

In [11]:
(df["children"].isnull().sum()*100)/(df.shape[0])


0.0033503643521232934

In [12]:
(df["country"].isnull().sum()*100)/(df.shape[0])


0.40874445095904177

In [13]:
(df["agent"].isnull().sum()*100)/(df.shape[0])


13.686238378423653

Filling the missing values of columns(children,agent,country) with fillna()method

In [14]:
df["children"].dtype

dtype('float64')

In [15]:
df["children"].median()

0.0

In [16]:
df["children"].fillna(df["children"].median(),inplace=True)

In [17]:
df["children"].isnull().sum()

0

In [18]:
df.country.dtype

dtype('O')

In [19]:
df.country.mode()

0    PRT
dtype: object

In [23]:
df.country.fillna("PRT",inplace=True)

In [24]:
df.country.isnull().sum()

0

In [25]:
df.agent.dtype

dtype('float64')

In [26]:
df.agent.median()

14.0

In [27]:
df.agent.fillna(df.agent.median(),inplace=True)

In [28]:
df.agent.isnull().sum()

0

In [29]:
df.isnull().sum()

hotel                             0
is_canceled                       0
lead_time                         0
arrival_date_year                 0
arrival_date_month                0
arrival_date_week_number          0
arrival_date_day_of_month         0
stays_in_weekend_nights           0
stays_in_week_nights              0
adults                            0
children                          0
babies                            0
meal                              0
country                           0
market_segment                    0
distribution_channel              0
is_repeated_guest                 0
previous_cancellations            0
previous_bookings_not_canceled    0
reserved_room_type                0
assigned_room_type                0
booking_changes                   0
deposit_type                      0
agent                             0
days_in_waiting_list              0
customer_type                     0
adr                               0
required_car_parking_spaces 

In [39]:
df["hotel"].unique()

array(['Resort Hotel', 'City Hotel'], dtype=object)

### Find Booking and Cancellation details of both hotels

In [32]:
matrix=df.groupby(["hotel"])["is_canceled"].value_counts()
matrix

hotel         is_canceled
City Hotel    0              46228
              1              33102
Resort Hotel  0              28938
              1              11122
Name: is_canceled, dtype: int64

1 indicates cancellation and 0 means not cancelled (active booking)

In [34]:
city_hotel=matrix[0]+matrix[1]
city_hotel

79330

In [35]:
resort_hotel=matrix[2]+matrix[3]
resort_hotel

40060

##### More bookings done in city Hotel than resort Hotel with following details:

total booking(city hotel) : 79330

Cancellation(city hotel) : 33102

active booking (city hotel) : 46228

total booking (resort hotel) : 40060

Cancellation(resort hotel) : 11122

active booking (resort hotel) : 28938

### Hotel city has more cancellation and more active bookings than resort hotel so we supposed that people visited City hotel

In [38]:
# Find booking and cancellation details throughout the year
df.groupby(["hotel","arrival_date_year"])["is_canceled"].value_counts()

hotel         arrival_date_year  is_canceled
City Hotel    2015               0               7678
                                 1               6004
              2016               0              22733
                                 1              15407
              2017               0              15817
                                 1              11691
Resort Hotel  2015               0               6176
                                 1               2138
              2016               0              13637
                                 1               4930
              2017               0               9125
                                 1               4054
Name: is_canceled, dtype: int64

#### maximum number of booking and cancellation done in 2016 in both the hotels 

In [42]:
# confirmed bookings
con_booking=df[df["is_canceled"]==0]
con_booking

hotel  is_canceled  lead_time  arrival_date_year  \
0       Resort Hotel            0        342               2015   
1       Resort Hotel            0        737               2015   
2       Resort Hotel            0          7               2015   
3       Resort Hotel            0         13               2015   
4       Resort Hotel            0         14               2015   
...              ...          ...        ...                ...   
119385    City Hotel            0         23               2017   
119386    City Hotel            0        102               2017   
119387    City Hotel            0         34               2017   
119388    City Hotel            0        109               2017   
119389    City Hotel            0        205               2017   

       arrival_date_month  arrival_date_week_number  \
0                    July                        27   
1                    July                        27   
2                    July                        27   
3                    July                        27   
4                    July                        27   
...                   ...                       ...   
119385             August                        35   
119386             August                        35   
119387             August                        35   
119388             August                        35   
119389             August                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
0                               1                        0   
1                               1                        0   
2                               1                        0   
3                               1                        0   
4                               1                        0   
...                           ...                      ...   
119385                         30                        2   
119386                         31                        2   
119387                         31                        2   
119388                         31                        2   
119389                         29                        2   

        stays_in_week_nights  adults  ...  booking_changes  deposit_type  \
0                          0       2  ...                3    No Deposit   
1                          0       2  ...                4    No Deposit   
2                          1       1  ...                0    No Deposit   
3                          1       1  ...                0    No Deposit   
4                          2       2  ...                0    No Deposit   
...                      ...     ...  ...              ...           ...   
119385                     5       2  ...                0    No Deposit   
119386                     5       3  ...                0    No Deposit   
119387                     5       2  ...                0    No Deposit   
119388                     5       2  ...                0    No Deposit   
119389                     7       2  ...                0    No Deposit   

        agent days_in_waiting_list customer_type     adr  \
0        14.0                    0     Transient    0.00   
1        14.0                    0     Transient    0.00   
2        14.0                    0     Transient   75.00   
3       304.0                    0     Transient   75.00   
4       240.0                    0     Transient   98.00   
...       ...                  ...           ...     ...   
119385  394.0                    0     Transient   96.14   
119386    9.0                    0     Transient  225.43   
119387    9.0                    0     Transient  157.71   
119388   89.0                    0     Transient  104.40   
119389    9.0                    0     Transient  151.20   

        required_car_parking_spaces  total_of_special_requests  \
0                                 0                          0   
1                                 0   

In [43]:
con_booking.hotel.unique()

array(['Resort Hotel', 'City Hotel'], dtype=object)

In [46]:
# now more details 
# How many bookings confirmed in both the hotels throughout the Year and month
con_booking.groupby(["hotel","arrival_date_year"])["arrival_date_month"].value_counts()

hotel         arrival_date_year  arrival_date_month
City Hotel    2015               October               2065
                                 September             1986
                                 August                1248
                                 December               986
                                 November               934
                                 July                   459
              2016               September             2304
                                 October               2272
                                 May                   2240
                                 June                  2203
                                 August                2131
                                 July                  2088
                                 April                 2022
                                 March                 1938
                                 November              1762
                                 February       

#### higest booking done in 

City ----> 2015 : oct , 2016 : sept ,2017 : may

resort ----> 2015 : oct ,2016 : oct ,2017 : may

In [48]:
# Find monthly guest booking trends at the city hotel
df1=con_booking[(con_booking["hotel"]=="City Hotel")]
arr=df1["arrival_date_month"].value_counts()

In [49]:
arr

August       5381
July         4782
May          4579
June         4366
October      4337
September    4290
March        4072
April        4015
February     3064
November     2696
December     2392
January      2254
Name: arrival_date_month, dtype: int64

largest month booking of city hotel : august


lowest month booking of city hotel : january

In [54]:
# Find monthly guest booking trends at the resort hotel
df2=con_booking[(con_booking["hotel"]=="Resort Hotel")]
arr=df2["arrival_date_month"].value_counts()

In [55]:
arr

August       3257
July         3137
October      2577
March        2573
April        2550
May          2535
February     2308
September    2102
June         2038
December     2017
November     1976
January      1868
Name: arrival_date_month, dtype: int64

largest month booking of city hotel : august

lowest month booking of city hotel : january

In [56]:
# Find Confirmed booking count of each category of Customers is as follows
con_booking.groupby(["hotel","customer_type"])["is_canceled"].value_counts()

hotel         customer_type    is_canceled
City Hotel    Contract         0               1195
              Group            0                264
              Transient        0              32306
              Transient-Party  0              12463
Resort Hotel  Contract         0               1619
              Group            0                254
              Transient        0              20793
              Transient-Party  0               6272
Name: is_canceled, dtype: int64

In [57]:
con_booking.groupby(["customer_type"])["is_canceled"].value_counts()

customer_type    is_canceled
Contract         0               2814
Group            0                518
Transient        0              53099
Transient-Party  0              18735
Name: is_canceled, dtype: int64

Transient type of customers had more bookings
Group type of customers had less bookings
so business managers target transient type of customers to increase references

In [58]:
#At What time transient type of customers visited the hotel
con_booking[con_booking["customer_type"]=="Transient"]["arrival_date_month"].mode()

0    August
dtype: object

In [61]:
#What type of room they booked
con_booking[con_booking["customer_type"]=="Transient"]["reserved_room_type"].mode()

0    A
dtype: object

In [62]:
#From which country they belong
con_booking[con_booking["customer_type"]=="Transient"]["country"].mode()

0    PRT
dtype: object

In [64]:
#Find Most demanded meal by guest
con_booking.groupby("hotel")["meal"].value_counts()

hotel         meal     
City Hotel    BB           35638
              SC            6601
              HB            3980
              FB               9
Resort Hotel  BB           22162
              HB            5499
              Undefined      883
              FB             311
              SC              83
Name: meal, dtype: int64

In [65]:
#Find Active booking count of each category of customers month wise
con_booking.groupby("customer_type")["arrival_date_month"].value_counts()

customer_type    arrival_date_month
Contract         September              595
                 October                459
                 August                 454
                 July                   290
                 June                   270
                 May                    246
                 November               147
                 April                  142
                 December               120
                 March                   42
                 February                34
                 January                 15
Group            October                 72
                 November                54
                 August                  50
                 February                47
                 December                44
                 May                     43
                 January                 40
                 March                   40
                 September               40
                 June                   

In [67]:
#From which country guest visited most and least
con_booking.groupby("hotel")["country"].value_counts()

hotel         country
City Hotel    PRT        10881
              FRA         7081
              DEU         5012
              GBR         3753
              ESP         3285
                         ...  
Resort Hotel  SYR            1
              TGO            1
              UGA            1
              UZB            1
              ZMB            1
Name: country, Length: 268, dtype: int64

In [68]:
len(con_booking[con_booking["reserved_room_type"]==con_booking["assigned_room_type"]]) *100 /len(con_booking["reserved_room_type"])

81.22156294069127

81% of guest got the same room type which they have booked

In [69]:
# What was longest and smallest strech of stays in weekend nights
con_booking.groupby("hotel")["stays_in_weekend_nights"].agg(["max","min"])

max  min
hotel                 
City Hotel     16    0
Resort Hotel   19    0

In [70]:
con_booking.groupby("hotel")["stays_in_week_nights"].agg(["max","min"])

max  min
hotel                 
City Hotel     41    0
Resort Hotel   50    0

### with respect to cancellation

In [71]:
df2=df[df["is_canceled"]==1]
df2

hotel  is_canceled  lead_time  arrival_date_year  \
8       Resort Hotel            1         85               2015   
9       Resort Hotel            1         75               2015   
10      Resort Hotel            1         23               2015   
27      Resort Hotel            1         60               2015   
32      Resort Hotel            1         96               2015   
...              ...          ...        ...                ...   
110280    City Hotel            1        132               2017   
111355    City Hotel            1          4               2017   
111924    City Hotel            1          7               2017   
111925    City Hotel            1          6               2017   
117295    City Hotel            1          0               2017   

       arrival_date_month  arrival_date_week_number  \
8                    July                        27   
9                    July                        27   
10                   July                        27   
27                   July                        27   
32                   July                        27   
...                   ...                       ...   
110280              April                        17   
111355               June                        23   
111924                May                        22   
111925               July                        29   
117295             August                        31   

        arrival_date_day_of_month  stays_in_weekend_nights  \
8                               1                        0   
9                               1                        0   
10                              1                        0   
27                              1                        2   
32                              1                        2   
...                           ...                      ...   
110280                         25                        0   
111355                          5                        1   
111924                         31                        0   
111925                         17                        1   
117295                          2                        0   

        stays_in_week_nights  adults  ...  booking_changes  deposit_type  \
8                          3       2  ...                0    No Deposit   
9                          3       2  ...                0    No Deposit   
10                         4       2  ...                0    No Deposit   
27                         5       2  ...                0    No Deposit   
32                         8       2  ...                0    No Deposit   
...                      ...     ...  ...              ...           ...   
110280                     0       0  ...                1    No Deposit   
111355                     0       1  ...                0    No Deposit   
111924                     1       1  ...                0    No Deposit   
111925                     0       1  ...                0    No Deposit   
117295                     2       1  ...                0    No Deposit   

        agent days_in_waiting_list customer_type    adr  \
8       240.0                    0     Transient   82.0   
9        15.0                    0     Transient  105.5   
10      240.0                    0     Transient  123.0   
27      240.0                    0     Transient  107.0   
32       14.0                    0     Transient  108.3   
...       ...                  ...           ...    ...   
110280    9.0                    0     Transient    0.0   
111355   14.0                    0     Transient   65.0   
111924   14.0                    0     Transient   65.0   
111925   14.0                    0     Transient   65.0   
117295   14.0                    0     Transient   65.0   

        required_car_parking_spaces  total_of_special_requests  \
8                                 0                          1   
9                                 0               

In [72]:
# how many cancellation done in each country
df2.groupby("hotel")["country"].value_counts()

hotel         country
City Hotel    PRT        20103
              FRA         1723
              GBR         1562
              ESP         1326
              ITA         1253
                         ...  
Resort Hotel  SEN            1
              SVN            1
              SYC            1
              TUN            1
              VEN            1
Name: country, Length: 196, dtype: int64

In [73]:
#Guest from PRT country did maximum cancellation
len(df2["country"])

44224

In [74]:
len(df2[(df2["country"]=="PRT")])

27586

In [76]:
can_per=27586*100/44224
can_per

62.37789435600579

 62% cancellation done from PRT country

In [77]:
#how many cancellation done in each year
df2.groupby("hotel")["arrival_date_year"].value_counts()

hotel         arrival_date_year
City Hotel    2016                 15407
              2017                 11691
              2015                  6004
Resort Hotel  2016                  4930
              2017                  4054
              2015                  2138
Name: arrival_date_year, dtype: int64

maximum cancellation done in 2016 on both the hotel